src/com/ansorgit/plugins/bash/lang/lexer/bash.flex

```diff
@@ -168,27 +168,34 @@ Filedescriptor = "&" {IntegerLiteral} | "&-"
 %xstate X_HEREDOC_MARKER
 %xstate X_HEREDOC_MARKER_IGNORE_TABS
 %xstate X_HEREDOC
+// this state is used to keep the lexer in a non-zero state when heredoc markers were found
+// if used as a highlighting lexer incremental lexing is used when in a zero state
+// heredoc lexing is stateful and is thus not compatible with the zero state
+%state S_HEREDOC_EXPECTED
 
 /* To match here-strings */
 %xstate X_HERE_STRING
 
 %%
 /***************************** INITIAL STAATE ************************************/
-<YYINITIAL, S_CASE, S_CASE_PATTERN, S_SUBSHELL, S_ASSIGNMENT_LIST> {
+<YYINITIAL, S_CASE, S_CASE_PATTERN, S_SUBSHELL, S_ASSIGNMENT_LIST, S_HEREDOC_EXPECTED> {
   {Shebang}                     { return SHEBANG; }
   {Comment}                     { return COMMENT; }
 }
 
 <X_HEREDOC_MARKER, X_HEREDOC_MARKER_IGNORE_TABS> {
+    // a line-feed after a heredoc marker indicates a missing start marker tag,
+    // the lexer must not remain in the heredoc marker state
+    {LineTerminator}             { heredocState().removeMarker(zzCurrentPos); backToPreviousState(); return LINE_FEED; }
     {WhiteSpaceLineCont}+        { return WHITESPACE; }
     {LineContinuation}+          { return WHITESPACE; }
-    {LineTerminator}             { return LINE_FEED; }
 
       ("$"? "'" [^\']+ "'")+
```

Comment:
Why not to use something like this
```
HeredocMarker            = [^\r\n|&\\;()[] \t\"'] | {EscapedChar}
HeredocMarkerInQuotes    = {HeredocMarker}+ | '{HeredocMarker}+' | \"{HeredocMarker}+\"
```
And here use `{HeredocMarkerInQuotes}`  it's more readable

Reply:
Good points. I noticed that the lexer is currently not properly supporting escaped characters. I'm tracking this at #741 .

I'm seeing these limitations for the rules above (I wasn't sure if this is from the Sh lexer or a draft):
- `'{HeredocMarker}+'` is accepting an escaped single quote, `'\''`, but escaped characters in single quotes are not allowed. 
- `$'\''` is accepting escaped characters, though. This isn't supported with the rules above.
- `{HeredocMarker}` is not matching space, but the quoted variants must accept space. This is valid Bash:
```bash
cat << "E O F"
content
E O F

cat << 'E O F'
content
E O F
```
- Sadly this this valid Bash and wouldn't be supported:
```bash
cat << $"EO""F"
content
EOF
```
- Sadly this is valid, too. This is not yet supported by this lexer, too.
```bash
cat << "E"O'F'
content
EOF
```